### PR TITLE
Refactor MySQL tests to use testcontainers

### PR DESCRIPTION
**Pull request by radical**
- Use ContainerImageTags.MySql.Tag as the server version
- Normalize connection strings with MySqlConnectionStringBuilder before comparing
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2906)

### CODE REVIEW

**Comment by radical** [comment]
Based on https://github.com/dotnet/aspire/pull/2862 . I will rebase this once that PR is merged.

**Comment by radical** [comment]
Rebased this on `main`, so it can be reviewed independently.

File: src/Aspire.Dashboard/Components/Controls/LogViewer.razor.css [hunk 1]
```diff
@@ -67,6 +67,7 @@
     user-select: text;
     cursor: text;
     white-space: pre-wrap;
+    overflow-wrap: anywhere;
```
Comment (eerhardt) [inline src/Aspire.Dashboard/Components/Controls/LogViewer.razor.css:70]: This looks unrelated. Can we revert it?

File: Directory.Packages.props [hunk 2]
```diff
@@ -134,6 +134,7 @@
     <PackageVersion Include="Testcontainers.RabbitMq" Version="$(TestcontainersPackageVersion)" />
     <PackageVersion Include="Testcontainers.MongoDb" Version="$(TestcontainersPackageVersion)" />
     <PackageVersion Include="Testcontainers.Redis" Version="$(TestcontainersPackageVersion)" />
+    <PackageVersion Include="Testcontainers.MySql" Version="$(TestcontainersPackageVersion)" />
```
Comment (eerhardt) [inline Directory.Packages.props:137]: (nit) it would be nice to keep these ordered.

File: tests/Aspire.MySqlConnector.Tests/ConformanceTests.cs [hunk 3]
```diff
@@ -114,45 +117,18 @@ public void BothDataSourceAndConnectionCanBeResolved(string? key)
...
-    [ConditionalFact]
-    public void TracingEnablesTheRightActivitySource_Keyed()
-    {
-        SkipIfCanNotConnectToServer();
-
-        RemoteExecutor.Invoke(() => ActivitySourceTest(key: "key")).Dispose();
-    }
+    [RequiresDockerFact]
+    public Task TracingEnablesTheRightActivitySource_Keyed()
+        => RunWithFixtureAsync(obj => obj.ActivitySourceTest(key: "key"));
```
Comment (eerhardt) [inline tests/Aspire.MySqlConnector.Tests/ConformanceTests.cs:126]: These tests still need to use RemoteExecutor.

File: tests/Aspire.Pomelo.EntityFrameworkCore.MySql.Tests/ConformanceTests.cs [hunk 4]
```diff
@@ -13,12 +13,13 @@
...
 {
-    // in the future it can become a static property that reads the value from Env Var
-    protected const string ConnectionString = "Server=localhost;User ID=root;Password=pass;Database=test";
-
-    private static readonly Lazy<bool> s_canConnectToServer = new(GetCanConnect);
+    private readonly MySqlContainerFixture _containerFixture;
+    protected string ConnectionString => RequiresDockerTheoryAttribute.IsSupported
+                                            ? _containerFixture.GetConnectionString()
+                                            : "Server=localhost;User ID=root;Password=pass;Database=test";
+    protected readonly string ServerVersion = $"{MySqlContainerImageTags.Tag}-{MySqlContainerImageTags.Image}";
```
Comment (eerhardt) [inline tests/Aspire.Pomelo.EntityFrameworkCore.MySql.Tests/ConformanceTests.cs:22]:
```suggestion
    protected readonly string ServerVersion = $"{MySqlContainerImageTags.Tag}-mysql";
```

This isn't the Image name. Yes they happen to be the same, but it's not what this value is. There are different flavors of MySql databases (ex. MariaDB). This string tells the EF library which kind of database to prepare statements for.

File: tests/Aspire.Pomelo.EntityFrameworkCore.MySql.Tests/ConformanceTests.cs [hunk 5]
```diff
@@ -119,28 +123,14 @@ public void DbContextCanBeAlwaysResolved()
...
-    [ConditionalFact]
-    public void TracingEnablesTheRightActivitySource()
-    {
-        SkipIfCanNotConnectToServer();
-
-        RemoteExecutor.Invoke(() => ActivitySourceTest(key: null)).Dispose();
-    }
+    [RequiresDockerFact]
+    public Task TracingEnablesTheRightActivitySource()
+        => RunWithFixtureAsync(obj => obj.ActivitySourceTest(key: null));
```
Comment (eerhardt) [inline tests/Aspire.Pomelo.EntityFrameworkCore.MySql.Tests/ConformanceTests.cs:128]: RemoteExecutor.

File: tests/Aspire.Pomelo.EntityFrameworkCore.MySql.Tests/AspireEFMySqlExtensionsTests.cs [hunk 6]
```diff
@@ -2,28 +2,38 @@
...
-public class AspireEFMySqlExtensionsTests
+public class AspireEFMySqlExtensionsTests : IClassFixture<MySqlContainerFixture>
 {
-    private const string ConnectionString = "Server=localhost;User ID=root;Database=test";
     private const string ConnectionStringSuffixAddedByPomelo = ";Allow User Variables=True;Use Affected Rows=False";
+    private readonly MySqlContainerFixture _containerFixture;
+    private string ConnectionString => RequiresDockerTheoryAttribute.IsSupported
+                                            ? _containerFixture.GetConnectionString()
+                                            : "Server=localhost;User ID=root;Password=pass;Database=test";
+    private readonly string _serverVersion = $"{MySqlContainerImageTags.Tag}-{MySqlContainerImageTags.Image}";
```
Comment (eerhardt) [inline tests/Aspire.Pomelo.EntityFrameworkCore.MySql.Tests/AspireEFMySqlExtensionsTests.cs:26]:
```suggestion
    private readonly string _serverVersion = $"{MySqlContainerImageTags.Tag}-mysql";
```

This isn't the Image name. Yes they happen to be the same, but it's not what this value is. There are different flavors of MySql databases (ex. MariaDB). This string tells the EF library which kind of database to prepare statements for.

**Review by eerhardt — COMMENTED** [review]
cc @bgrainger

File: tests/Aspire.MySqlConnector.Tests/ConformanceTests.cs [hunk 7]
```diff
@@ -12,12 +13,12 @@
...
 
-public class ConformanceTests : ConformanceTests<MySqlDataSource, MySqlConnectorSettings>
+public class ConformanceTests : ConformanceTests<MySqlDataSource, MySqlConnectorSettings>, IClassFixture<MySqlContainerFixture>
 {
-    private const string ConnectionSting = "Host=localhost;Database=test_aspire_mysql;Username=root;Password=password";
-
-    private static readonly Lazy<bool> s_canConnectToServer = new(GetCanConnect);
-
+    private readonly MySqlContainerFixture _containerFixture;
+    private string ConnectionSting => RequiresDockerTheoryAttribute.IsSupported
```
Comment (bgrainger) [inline tests/Aspire.MySqlConnector.Tests/ConformanceTests.cs:19]:
```suggestion
    private string ConnectionString => RequiresDockerTheoryAttribute.IsSupported
```

File: tests/Aspire.MySqlConnector.Tests/ConformanceTests.cs [hunk 8]
```diff
@@ -12,12 +13,12 @@
...
+                                        ? _containerFixture.GetConnectionString()
+                                        : "Host=localhost;Database=test_aspire_mysql;Username=root;Password=password";
```
Comment (bgrainger) [inline tests/Aspire.MySqlConnector.Tests/ConformanceTests.cs:21]:
```suggestion
                                        : "Server=localhost;User ID=root;Password=password;Database=test_aspire_mysql";
```

Don't know if it matters, but the "normalized" version of a MySqlConnector connection string would use different keywords (`Server` not `Host`) and a slightly different order.

File: tests/Aspire.MySqlConnector.Tests/ConformanceTests.cs (hunk 7 above)
Comment (radical) [reply tests/Aspire.MySqlConnector.Tests/ConformanceTests.cs:19]: good catch!

File: tests/Aspire.MySqlConnector.Tests/ConformanceTests.cs (hunk 8 above)
Comment (bgrainger) [inline tests/Aspire.MySqlConnector.Tests/ConformanceTests.cs:21]: This `private` property doesn't appear to be used; is it necessary?

Comment (radical) [reply tests/Aspire.MySqlConnector.Tests/ConformanceTests.cs:21]: It is used in `PopulateConfiguration`.

Comment (radical) [reply tests/Aspire.MySqlConnector.Tests/ConformanceTests.cs:21]: Indeed, I was hitting that - https://github.com/dotnet/aspire/commit/491a14b473964e06ef458273d7fe43d5bd43df88

Comment (bgrainger) [reply tests/Aspire.MySqlConnector.Tests/ConformanceTests.cs]: Right you are; sorry I missed that when reading through the file.

File: src/Aspire.Dashboard/Components/Controls/LogViewer.razor.css (hunk 1 above)
Comment (eerhardt) [reply src/Aspire.Dashboard/Components/Controls/LogViewer.razor.css:70]: This doesn't appear to be reverted yet.

File: tests/Aspire.MySqlConnector.Tests/MySqlContainerFixture.cs [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+    public string GetConnectionString() => Container?.GetConnectionString() ??
+        throw new InvalidOperationException("The test container was not initialized.");
+
+    public async Task InitializeAsync()
+    {
+        if (RequiresDockerTheoryAttribute.IsSupported)
+        {
+            // testcontainers uses mongo:mongo by default,
+            // resetting that for tests
```
Comment (eerhardt) [inline tests/Aspire.MySqlConnector.Tests/MySqlContainerFixture.cs:23]:
```suggestion
```

This doesn't add value.

File: tests/Aspire.Pomelo.EntityFrameworkCore.MySql.Tests/ConformanceTests.cs (hunk 5 above)
Comment (eerhardt) [inline tests/Aspire.Pomelo.EntityFrameworkCore.MySql.Tests/ConformanceTests.cs:128]:
```suggestion
        => RemoteExecutor.Invoke(() => RunWithFixtureAsync(obj => obj.ActivitySourceTest(key: null))).Dispose();
```

File: tests/Aspire.Pomelo.EntityFrameworkCore.MySql.Tests/AspireEFMySqlExtensionsTests.cs [hunk 10]
```diff
@@ -81,14 +95,14 @@ public void CanConfigureDbContextOptions()
...
-            new KeyValuePair<string, string?>("Aspire:Pomelo:EntityFrameworkCore:MySql:ServerVersion", "8.2.0-mysql"),
+            new KeyValuePair<string, string?>("Aspire:Pomelo:EntityFrameworkCore:MySql:ServerVersion", _serverVersion),
             new KeyValuePair<string, string?>("ConnectionStrings:mysql", ConnectionString),
             new KeyValuePair<string, string?>("Aspire:Pomelo:EntityFrameworkCore:MySql:Retry", "true")
         ]);
 
         builder.AddMySqlDbContext<TestDbContext>("mysql", configureDbContextOptions: optionsBuilder =>
         {
-            optionsBuilder.UseMySql(new MySqlServerVersion(new Version(8, 2, 0)), mySqlBuilder =>
+            optionsBuilder.UseMySql(new MySqlServerVersion(new Version(MySqlContainerImageTags.Tag)), mySqlBuilder =>
```
Comment (eerhardt) [inline tests/Aspire.Pomelo.EntityFrameworkCore.MySql.Tests/AspireEFMySqlExtensionsTests.cs:105]: It might make sense to pull `new MySqlServerVersion(new Version(MySqlContainerImageTags.Tag))` into a single location so all these tests don't need to duplicate it.

File: tests/Aspire.Hosting.Tests/MySql/AddMySqlTests.cs [hunk 11]
```diff
@@ -2,10 +2,10 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using System.Net.Sockets;
-using Aspire.Hosting.MySql;
 using System.Text.RegularExpressions;
 using Microsoft.Extensions.DependencyInjection;
 using Xunit;
+using Aspire.Hosting.MySql;
 using Aspire.Hosting.Utils;
 using Aspire.Hosting.Tests.Utils;
```
Comment (bgrainger) [inline tests/Aspire.Hosting.Tests/MySql/AddMySqlTests.cs:10]: Nitpick: based on other files (and general .NET convention) `using Aspire` should come before `using Microsoft` (but after `using System`) due to alphabetical ordering.

File: tests/Aspire.MySqlConnector.Tests/AspireMySqlConnectorExtensionsTests.cs [hunk 12]
```diff
@@ -37,7 +44,9 @@ public void ReadsFromConnectionStringsCorrectly(bool useKeyed)
             host.Services.GetRequiredKeyedService<MySqlDataSource>("mysql") :
             host.Services.GetRequiredService<MySqlDataSource>();
 
-        Assert.Equal(ConnectionString, dataSource.ConnectionString);
+        string actualConnectionString = dataSource.ConnectionString;
+        string expectedConnectionString = new MySqlConnectionStringBuilder(ConnectionString).ConnectionString;
```
Comment (bgrainger) [inline tests/Aspire.MySqlConnector.Tests/AspireMySqlConnectorExtensionsTests.cs:48]:
```suggestion
        string expectedConnectionString = ConnectionString;
```

I think this should be unnecessary? The `ConnectionString` constant is already in the "normalized" form so this round-tripping through `MySqlConnectionStringBuilder` will be a no-op.

Comment (bgrainger) [inline tests/Aspire.MySqlConnector.Tests/AspireMySqlConnectorExtensionsTests.cs:48]: `new MySqlConnectionStringBuilder(ConnectionString).ConnectionString` is repeated a few times in this file; it might be worth creating a `private string NormalizedConnectionString => new MySqlConnectionStringBuilder(ConnectionString).ConnectionString` property to encapsulate it?

File: tests/Aspire.MySqlConnector.Tests/AspireMySqlConnectorExtensionsTests.cs [hunk 13]
```diff
@@ -37,7 +44,9 @@ public void ReadsFromConnectionStringsCorrectly(bool useKeyed)
...
+        Assert.Equal(expectedConnectionString, actualConnectionString);
```
Comment (bgrainger) [inline tests/Aspire.MySqlConnector.Tests/AspireMySqlConnectorExtensionsTests.cs:49]:
```suggestion
        Assert.Equal(NormalizedConnectionString, dataSource.ConnectionString);
```

With that new property, the assertion (here, and repeated below) could be simplified.

File: tests/Aspire.Hosting.Tests/MySql/AddMySqlTests.cs [hunk 14]
```diff
@@ -1,13 +1,13 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using Aspire.Hosting.MySql;
+using Aspire.Hosting.Utils;
+using Aspire.Hosting.Tests.Utils;
 using System.Net.Sockets;
 using System.Text.RegularExpressions;
```
Comment (bgrainger) [inline tests/Aspire.Hosting.Tests/MySql/AddMySqlTests.cs:8]: `using System` still comes first (conventionally).

Comment (bgrainger) [inline tests/Aspire.Hosting.Tests/MySql/AddMySqlTests.cs:8]:
```suggestion
using System.Net.Sockets;
using System.Text.RegularExpressions;
using Aspire.Hosting.MySql;
using Aspire.Hosting.Tests.Utils;
using Aspire.Hosting.Utils;
```

Others should be alphabetical.

File: tests/Aspire.Pomelo.EntityFrameworkCore.MySql.Tests/ConformanceTests.cs (hunk 5 above)
Comment (eerhardt) [reply tests/Aspire.Pomelo.EntityFrameworkCore.MySql.Tests/ConformanceTests.cs:128]: Has this been addressed?

Comment (radical) [reply tests/Aspire.Pomelo.EntityFrameworkCore.MySql.Tests/ConformanceTests.cs:128]: Sorry! I kept looking at the other instance of the same thing. Fixing now.

**Review by eerhardt — APPROVED** [review]
LGTM. Thanks!

**Review by bgrainger — APPROVED** [review]
